src/tar_parser.rs

```diff
@@ -0,0 +1,246 @@
+//! A This module implements a parser for uncompressed tar files,
+//! including the ustar extension (but not yet supporting the PAX extension).
+
+use core::ops::Range;
+
+use crate::filename::{truncate, Filename};
+
+/// A parser for an already decompressed image / tar ball
+#[derive(Clone, Copy)]
+pub struct Parser<'a> {
+    input: &'a BytesSlice,
+    offset: usize,
+}
+
+impl<'a> Parser<'a> {
+    /// Constructs a new parser from a given tar byte slice
+    pub fn new(input: &'a BytesSlice) -> Self {
+        Self { input, offset: 0 }
+    }
+}
+
+/// Metadata for a tar file entry.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+pub struct FileMetadata {
+    /// Is the file entry marked as executable?
+    pub is_exec: bool,
+}
+
+/// A file entry in a tar ball
+#[derive(Clone, Debug)]
+pub struct File<'a> {
+    /// The full file name of the entry
+    pub name: Filename<'a>,
+
+    /// The associated file metadata
+    pub metadata: FileMetadata,
+
+    /// The indices range of the value associated to this entry
+    pub value_range: Range<usize>,
+
+    /// A slice of the value (file content) associated to this entry
+    pub value: &'a [u8],
+}
+
+/// Possible parser errors:
+#[derive(Clone, Debug, PartialEq)]
+pub enum ParserError<'a> {
+    /// The parser encountered the end of the file during parsing of an entry.
+    UnexpectedEof,
+
+    /// The parser encountered something which was supposed to be an integer, but wasn't.
+    ParseInt(core::num::ParseIntError),
+
+    /// The parser encountered an integer which was out-of-range.
+    FromInt(core::num::TryFromIntError),
+
+    /// The parser encountered something which
+    /// was supposed to be an UTF-8 string (usually an ASCII integer), but wasn't.
+    Utf8(core::str::Utf8Error),
+
+    /// The thin tree encountered a file which got overridden by a directory.
+    FileOverridenWithDirectory(Filename<'a>),
+}
+
+impl<'a> From<core::num::ParseIntError> for ParserError<'a> {
+    #[inline(always)]
+    fn from(x: core::num::ParseIntError) -> Self {
+        Self::ParseInt(x)
+    }
+}
+
+impl<'a> From<core::num::TryFromIntError> for ParserError<'a> {
+    #[inline(always)]
+    fn from(x: core::num::TryFromIntError) -> Self {
+        Self::FromInt(x)
+    }
+}
+
+impl<'a> From<core::str::Utf8Error> for ParserError<'a> {
+    #[inline(always)]
+    fn from(x: core::str::Utf8Error) -> Self {
+        Self::Utf8(x)
+    }
+}
+
+#[cfg(feature = "alloc")]
+/// Owned bytes belonging to a decompressed tar file
+#[derive(Clone, Default, PartialEq, Eq)]
+pub struct Bytes(pub alloc::boxed::Box<[u8]>);
```

Comment:
I think, instead of having two Bytes structs, you can use a Trait:
```rs
/// Data representing the content of a Tar File
pub struct TarBytes<S: AsRef<[u8]>>(pub S);

fn main() {
    let b = Box::new([1_u8,2,3]);
    let a = [1_u8,2,3];
    let v = vec!(1_u8,2,3,4);
    let _t = TarBytes(&a);
    let _t = TarBytes(a);
    let _t = TarBytes(*b);
    let _t = TarBytes(&v);
    let _t = TarBytes(v);
}
```